File: sdk/swift/Tests/SeocheonSDKTests/Infrastructure/EnvelopeTests.swift
import XCTest
@testable import SeocheonSDK

final class EnvelopeTests: XCTestCase {

    func testEncodeTxBody() {
        let msg = MsgSubmitActivity(submitter: "seocheon1abc", activityHash: String(repeating: "a", count: 64), contentURI: "ipfs://test")
        let body = Envelope.encodeTxBody(messages: [msg], memo: "", timeoutHeight: 0)
        XCTAssertFalse(body.isEmpty)
    }

    func testEncodeAuthInfo() {
        let pubKey = Data(repeating: 0x02, count: 33)
        let coins = [Coin(denom: "uppyeo", amount: "50000")]
        let authInfo = Envelope.encodeAuthInfo(pubKey: pubKey, sequence: 0, feeCoins: coins, gasLimit: 200000)
        XCTAssertFalse(authInfo.isEmpty)
    }

    func testEncodeSignDoc() {
        let bodyBytes = Data([0x01, 0x02])
        let authInfoBytes = Data([0x03, 0x04])
        let signDoc = Envelope.encodeSignDoc(bodyBytes: bodyBytes, authInfoBytes: authInfoBytes, chainID: "seocheon-1", accountNumber: 1)
        XCTAssertFalse(signDoc.isEmpty)
    }

    func testEncodeTxRaw() {
        let bodyBytes = Data([0x01, 0x02])
        let authInfoBytes = Data([0x03, 0x04])
        let signature = Data(repeating: 0xAA, count: 64)
        let txRaw = Envelope.encodeTxRaw(bodyBytes: bodyBytes, authInfoBytes: authInfoBytes, signatures: [signature])
        XCTAssertFalse(txRaw.isEmpty)
    }

    func testEncodeAny() {
        let data = Data([0x01])
        let anyData = Envelope.encodeAny(typeURL: "/test.Type", value: data)
        XCTAssertFalse(anyData.isEmpty)
    }

    func testEncodePubKeyAny() {
        let pubKey = Data(repeating: 0x02, count: 33)
        let pubKeyAny = Envelope.encodePubKeyAny(pubKey)
        XCTAssertFalse(pubKeyAny.isEmpty)
    }
}

File: sdk/csharp/tests/Internal/Tx/EnvelopeTests.cs
using Seocheon.Sdk.Internal.Tx;
using Xunit;

namespace Seocheon.Sdk.Tests.Internal.Tx;

public class EnvelopeTests
{
    [Fact]
    public void EncodeAny_ContainsTypeUrl()
    {
        var any = Envelope.EncodeAny("/test.Type", [0x01, 0x02]);
        Assert.NotEmpty(any);
        Assert.Contains((byte)0x0A, any); // Field 1 (type_url)
    }

    [Fact]
    public void EncodeTxBody_WithMessage()
    {
        var msgAny = Envelope.EncodeAny("/test.Msg", [0x01]);
        var body = Envelope.EncodeTxBody(msgAny);
        Assert.NotEmpty(body);
    }

    [Fact]
    public void EncodeTxBody_WithMemo()
    {
        var msgAny = Envelope.EncodeAny("/test.Msg", [0x01]);
        var body = Envelope.EncodeTxBody(msgAny, memo: "test memo");
        Assert.NotEmpty(body);
        // Should be larger than without memo
        var bodyNoMemo = Envelope.EncodeTxBody(msgAny);
        Assert.True(body.Length > bodyNoMemo.Length);
    }

    [Fact]
    public void EncodeAuthInfo_NotEmpty()
    {
        var pubKey = new byte[33];
        pubKey[0] = 0x02;
        var authInfo = Envelope.EncodeAuthInfo(pubKey, 0, 200000, 50000000, "uppyeo");
        Assert.NotEmpty(authInfo);
    }

    [Fact]
    public void EncodeSignDoc_NotEmpty()
    {
        var signDoc = Envelope.EncodeSignDoc(
            [0x01], // body
            [0x02], // auth_info
            "seocheon-1",
            42
        );
        Assert.NotEmpty(signDoc);
    }

    [Fact]
    public void EncodeTxRaw_NotEmpty()
    {
        var txRaw = Envelope.EncodeTxRaw(
            [0x01],       // body
            [0x02],       // auth_info
            new byte[64]  // signature
        );
        Assert.NotEmpty(txRaw);
    }
}
